File: Sorting/Quick_sort/quick_sort.py
# Quick Sort implementation using python

def Partition(A, p, r):
    pivot = A[r]
    i = p - 1
    for j in range(p, r):
        if (A[j] <= pivot):
            i = i + 1
            A[i], A[j] = A[j], A[i]
    
    A[i+1], A[r] = A[r], A[i+1]
    return i+1

def QuickSort(A, p, r):
    q = Partition(A, p, r)
    if (p < q):
        QuickSort(A, p, q-1)
        QuickSort(A, q+1, r)

def printArray(A):
    for i in range(len(A)):
        print(A[i], end=" ")
    print()

if __name__ == "__main__":
    A = [5, 2, 4, 6, 1, 3] 
    print("Unsorted Array: ", end="")
    printArray(A)
    QuickSort(A, 0, len(A)-1)
    print("Sorted Array: ", end="")
    printArray(A)

File: Sorting/quick_sort.c
#include <stdio.h>

/*Quick sort algorithm for sorting
 Time complexity of Quicksort :
    - Best case :- O(nlogn)
    - Worst case :- O(n^2)
    - Average case :- O(nlogn)
*/

int Partition(int A[], int p, int r)
{
    int x, i, j, temp;
    x = A[r];
    i = p - 1;
    for (j = p; j <= r - 1; j++)
    {
        if (A[j] <= x)
        {
            i = i + 1;
            temp = A[i];
            A[i] = A[j];
            A[j] = temp;
        }
    }
    temp = A[i + 1];
    return i+1;
}

int QuickSort(int A[], int p, int r)
{
    int q;
    if (p < r)
    {
        q = Partition(A, p, r);
        QuickSort(A, p, q - 1);
        QuickSort(A, q + 1, r);
    }
    return A;
}

int main()
{   
    int A = [9,10,5,3,2,1,4,8,7,6];
    int i;
    QuickSort(A, 0, 9);
    printf("Sorted array is \n");
    for (i = 0; i <= 9; i++)
        printf("%d ", A[i]);
    return 0;

}
